src/expr/src/relation/func.rs

```diff
@@ -453,48 +458,98 @@ where
-                    let key = list.next().unwrap();
-                    let val = list.next().unwrap();
-                    if key.is_null() {
-                        // TODO(benesch): this should produce an error, but
-                        // aggregate functions cannot presently produce errors.
-                        None
-                    } else {
-                        Some((key.unwrap_str(), val))
-                    }
-                })
-                .sorted_by_key(|(k, _v)| *k)
-                .dedup_by(|(k1, _v1), (k2, _v2)| k1 == k2),
-        );
+        let mut datums: Vec<_> = datums
+            .into_iter()
+            .filter_map(|d| {
+                if d.is_null() {
+                    return None;
+                }
+                let mut list = d.unwrap_list().iter();
+                let key = list.next().unwrap();
+                let val = list.next().unwrap();
+                if key.is_null() {
+                    // TODO(benesch): this should produce an error, but
+                    // aggregate functions cannot presently produce errors.
+                    None
+                } else {
+                    Some((key.unwrap_str(), val))
+                }
+            })
+            .collect();
+        // datums are ordered by any ORDER BY clause now, and we want to preserve the
+        // last entry for each key, but we also need to present unique and sorted keys
+        // to present_dict. Use sort_by here, which is stable, and so will preserve the
+        // ORDER BY order. Then reverse and dedup to retain the last of each
+        // key. Reverse again so we're back in push_dict order.
+        datums.sort_by_key(|(k, _v)| *k);
+        datums.reverse();
+        datums.dedup_by_key(|(k, _v)| *k);
+        datums.reverse();
+        packer.push_dict(datums);
     })
 }
 
-fn array_concat<'a, I>(datums: I, temp_storage: &'a RowArena) -> Datum<'a>
+fn order_aggregate_datums<'a, I>(
+    datums: I,
+    order_by: &Vec<ColumnOrder>,
+) -> impl Iterator<Item = Datum<'a>>
 where
     I: IntoIterator<Item = Datum<'a>>,
 {
+    let mut rows: Vec<(Datum, Row)> = datums
+        .into_iter()
+        .filter_map(|d| {
+            // Filter out the identity datum.
+            if d.is_null() {
+                None
+            } else {
+                let list = d.unwrap_list();
+                let expr = list.iter().next().unwrap();
+                let order_row = Row::pack(list.iter().skip(1));
+                Some((expr, order_row))
+            }
+        })
+        .collect();
+    let mut sort_by = |left: &(_, Row), right: &(_, Row)| {
+        let left = &left.1;
+        let right = &right.1;
+        compare_columns(&order_by, &left.unpack(), &right.unpack(), || {
+            left.cmp(&right)
+        })
+    };
+    rows.sort_by(&mut sort_by);
+    rows.into_iter().map(|(expr, _order_row)| expr)
+}
+
+fn array_concat<'a, I>(
```

Comment:
Ah, `array_concat` is here now. No comment needed here, but one up above would be great!